File: modules/agent/src/services/project_service.py
"""Service layer for project business logic."""

import logging
from datetime import date
from typing import Optional, Dict, Any, List

from fastapi import HTTPException

from repositories.project_repository import (
    find_all_projects,
    find_project_by_id,
    count_project_deals,
    count_project_leads,
    create_project as create_project_repo,
    update_project as update_project_repo,
    delete_project as delete_project_repo,
    find_project_leads,
    find_project_deals,
    ProjectCreate,
    ProjectUpdate,
)

# Re-export Pydantic models for controllers
__all__ = ["ProjectCreate", "ProjectUpdate"]

logger = logging.getLogger(__name__)


def _parse_date(date_str: Optional[str]) -> Optional[date]:
    """Parse date string to date object."""
    if not date_str:
        return None
    return date.fromisoformat(date_str[:10])


async def get_projects(tenant_id: Optional[int]) -> List:
    """Get all projects for tenant."""
    return await find_all_projects(tenant_id)


async def get_project(project_id: int, tenant_id: Optional[int]):
    """Get project by ID with related counts."""
    project = await find_project_by_id(project_id, tenant_id)

    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    deals_count = await count_project_deals(project_id)
    leads_count = await count_project_leads(project_id)

    return project, deals_count, leads_count


async def create_project(
    project_data: Dict[str, Any],
    tenant_id: Optional[int],
    user_id: Optional[int],
):
    """Create a new project."""
    return await create_project_repo(
        tenant_id=tenant_id,
        name=project_data.get("name"),
        description=project_data.get("description"),
        status=project_data.get("status"),
        current_status_id=project_data.get("current_status_id"),
        start_date=_parse_date(project_data.get("start_date")),
        end_date=_parse_date(project_data.get("end_date")),
        user_id=user_id,
    )


async def update_project(
    project_id: int,
    project_data: Dict[str, Any],
    tenant_id: Optional[int],
    user_id: Optional[int],
):
    """Update a project."""
    # Parse dates before passing to repository
    data = dict(project_data)
    if data.get("start_date") is not None:
        data["start_date"] = _parse_date(data["start_date"])
    if data.get("end_date") is not None:
        data["end_date"] = _parse_date(data["end_date"])

    project = await update_project_repo(project_id, tenant_id, data, user_id)

    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    return project


async def delete_project(project_id: int, tenant_id: Optional[int]):
    """Delete a project."""
    deleted = await delete_project_repo(project_id, tenant_id)

    if not deleted:
        raise HTTPException(status_code=404, detail="Project not found")

    return True


async def get_project_leads(project_id: int, tenant_id: Optional[int]) -> List:
    """Get leads for a project."""
    return await find_project_leads(project_id, tenant_id)


async def get_project_deals(project_id: int, tenant_id: Optional[int]) -> List:
    """Get deals for a project."""
    return await find_project_deals(project_id, tenant_id)

File: modules/agent-go/internal/services/project_service.go
package services

import (
	"errors"
	"time"

	"github.com/pina-colada-co/agent-go/internal/models"
	"github.com/pina-colada-co/agent-go/internal/repositories"
	"github.com/pina-colada-co/agent-go/internal/schemas"
)

var errProjectNotFound = errors.New("project not found")

// ProjectService handles project business logic
type ProjectService struct {
	projectRepo *repositories.ProjectRepository
}

// NewProjectService creates a new project service
func NewProjectService(projectRepo *repositories.ProjectRepository) *ProjectService {
	return &ProjectService{projectRepo: projectRepo}
}

// ProjectResponse represents a project in API responses
type ProjectResponse struct {
	ID          int64   `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description"`
	Status      *string `json:"status"`
}

// ProjectDetailResponse represents a project with counts
type ProjectDetailResponse struct {
	ID          int64   `json:"id"`
	Name        string  `json:"name"`
	Description *string `json:"description"`
	Status      *string `json:"status"`
	DealsCount  int64   `json:"deals_count"`
	LeadsCount  int64   `json:"leads_count"`
}

// LeadResponse represents a lead in API responses
type LeadResponse struct {
	ID       int64   `json:"id"`
	DealID   int64   `json:"deal_id"`
	Type     string  `json:"type"`
	Source   *string `json:"source"`
}

// GetProjects returns all projects for a tenant
func (s *ProjectService) GetProjects(tenantID *int64) ([]ProjectResponse, error) {
	projects, err := s.projectRepo.FindAll(tenantID)
	if err != nil {
		return nil, err
	}

	results := make([]ProjectResponse, len(projects))
	for i, p := range projects {
		results[i] = ProjectResponse{
			ID:          p.ID,
			Name:        p.Name,
			Description: p.Description,
			Status:      p.Status,
		}
	}

	return results, nil
}

// GetProject returns a project by ID with counts
func (s *ProjectService) GetProject(id int64, tenantID *int64) (*ProjectDetailResponse, error) {
	project, err := s.projectRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if project == nil {
		return nil, errProjectNotFound
	}

	// Verify tenant if provided
	if tenantID != nil && project.TenantID != nil && *project.TenantID != *tenantID {
		return nil, errProjectNotFound
	}

	dealsCount, _ := s.projectRepo.GetDealsCount(id)
	leadsCount, _ := s.projectRepo.GetLeadsCount(id)

	return &ProjectDetailResponse{
		ID:          project.ID,
		Name:        project.Name,
		Description: project.Description,
		Status:      project.Status,
		DealsCount:  dealsCount,
		LeadsCount:  leadsCount,
	}, nil
}

// CreateProject creates a new project
func (s *ProjectService) CreateProject(input schemas.ProjectCreate, tenantID *int64, userID int64) (*ProjectResponse, error) {
	project := &models.Project{
		TenantID:    tenantID,
		Name:        input.Name,
		Description: input.Description,
		Status:      input.Status,
		CreatedBy:   userID,
		UpdatedBy:   userID,
	}

	if input.CurrentStatusID != nil {
		project.CurrentStatusID = input.CurrentStatusID
	}
	if input.StartDate != nil {
		if t, err := time.Parse("2006-01-02", *input.StartDate); err == nil {
			project.StartDate = &t
		}
	}
	if input.EndDate != nil {
		if t, err := time.Parse("2006-01-02", *input.EndDate); err == nil {
			project.EndDate = &t
		}
	}

	if err := s.projectRepo.Create(project); err != nil {
		return nil, err
	}

	return &ProjectResponse{
		ID:          project.ID,
		Name:        project.Name,
		Description: project.Description,
		Status:      project.Status,
	}, nil
}

// UpdateProject updates a project
func (s *ProjectService) UpdateProject(id int64, input schemas.ProjectUpdate, tenantID *int64, userID int64) (*ProjectResponse, error) {
	project, err := s.projectRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if project == nil {
		return nil, errProjectNotFound
	}

	// Verify tenant if provided
	if tenantID != nil && project.TenantID != nil && *project.TenantID != *tenantID {
		return nil, errProjectNotFound
	}

	updates := buildProjectUpdates(input, userID)
	if len(updates) > 0 {
		if err := s.projectRepo.Update(id, updates); err != nil {
			return nil, err
		}
	}

	// Refetch
	project, _ = s.projectRepo.FindByID(id)
	return &ProjectResponse{
		ID:          project.ID,
		Name:        project.Name,
		Description: project.Description,
		Status:      project.Status,
	}, nil
}

func buildProjectUpdates(input schemas.ProjectUpdate, userID int64) map[string]interface{} {
	updates := map[string]interface{}{"updated_by": userID}
	if input.Name != nil {
		updates["name"] = *input.Name
	}
	if input.Description != nil {
		updates["description"] = *input.Description
	}
	if input.Status != nil {
		updates["status"] = *input.Status
	}
	if input.CurrentStatusID != nil {
		updates["current_status_id"] = *input.CurrentStatusID
	}
	if input.StartDate != nil {
		if t, err := time.Parse("2006-01-02", *input.StartDate); err == nil {
			updates["start_date"] = t
		}
	}
	if input.EndDate != nil {
		if t, err := time.Parse("2006-01-02", *input.EndDate); err == nil {
			updates["end_date"] = t
		}
	}
	return updates
}

// DeleteProject deletes a project
func (s *ProjectService) DeleteProject(id int64, tenantID *int64) error {
	project, err := s.projectRepo.FindByID(id)
	if err != nil {
		return err
	}
	if project == nil {
		return errProjectNotFound
	}

	// Verify tenant if provided
	if tenantID != nil && project.TenantID != nil && *project.TenantID != *tenantID {
		return errProjectNotFound
	}

	return s.projectRepo.Delete(id)
}

// GetProjectLeads returns leads for a project
func (s *ProjectService) GetProjectLeads(projectID int64, tenantID *int64) ([]LeadResponse, error) {
	leads, err := s.projectRepo.GetProjectLeads(projectID, tenantID)
	if err != nil {
		return nil, err
	}

	results := make([]LeadResponse, len(leads))
	for i, l := range leads {
		results[i] = LeadResponse{
			ID:     l.ID,
			DealID: l.DealID,
			Type:   l.Type,
			Source: l.Source,
		}
	}
	return results, nil
}

// GetProjectDeals returns deals for a project
func (s *ProjectService) GetProjectDeals(projectID int64, tenantID *int64) ([]LeadResponse, error) {
	deals, err := s.projectRepo.GetProjectDeals(projectID, tenantID)
	if err != nil {
		return nil, err
	}

	results := make([]LeadResponse, len(deals))
	for i, d := range deals {
		results[i] = LeadResponse{
			ID:     d.ID,
			DealID: d.DealID,
			Type:   d.Type,
			Source: d.Source,
		}
	}
	return results, nil
}
